core/rpc/core/src/impl/utils.rs

```diff
@@ -1449,116 +1468,129 @@ impl<C: CkbRpc> MercuryRpcImpl<C> {
+                    )
+                    .await;
+                    return Ok(Some(transfer_components.outputs.len() - 1));
                 }
             }
             Some(ref change_address) => {
                 // change to tx outputs cell with same address
                 for (index, output_cell) in transfer_components.outputs.iter_mut().enumerate() {
                     let cell_address = self.script_to_address(&output_cell.lock()).to_string();
                     if *change_address == cell_address {
-                        let current_capacity: u64 = output_cell.capacity().unpack();
-                        let new_capacity =
-                            u64::try_from(current_capacity as i128 - required_capacity)
-                                .expect("impossible: overflow");
-                        let new_output_cell = output_cell
-                            .clone()
-                            .as_builder()
-                            .capacity(new_capacity.pack())
-                            .build();
-                        *output_cell = new_output_cell;
-                        if pay_fee.is_some() {
-                            transfer_components.fee_change_cell_index = Some(index);
-                        }
-                        return Ok(());
+                        change_to_existed_cell(output_cell, change_capacity);
+                        return Ok(Some(index));
                     }
                 }
             }
         }
 
-        // change new secp ckb cell to output, may need new live cells
-        if required_capacity.unsigned_abs() < MIN_CKB_CAPACITY as u128 {
-            loop {
-                if required_capacity.unsigned_abs() >= MIN_CKB_CAPACITY as u128 {
-                    break;
-                }
-                let (live_cell, asset_script_type) = self
-                    .pool_next_live_cell_for_capacity(
-                        ctx.clone(),
-                        &mut ckb_cells_cache,
-                        required_capacity,
-                    )
-                    .await?;
-                if self.is_in_cache(&live_cell.out_point) {
-                    continue;
-                }
-                if input_cell_set.contains(&live_cell.out_point) {
-                    continue;
-                }
-                input_cell_set.insert(live_cell.out_point.clone());
-                let capacity_provided = self
-                    .add_live_cell_for_balance_capacity(
-                        ctx.clone(),
-                        live_cell,
-                        asset_script_type,
-                        required_capacity,
-                        transfer_components,
-                    )
-                    .await;
-                required_capacity -= capacity_provided as i128;
+        Ok(None)
+    }
+
+    fn find_acp_or_secp_belong_to_items<'a>(
+        &self,
+        cells: &'a mut Vec<packed::CellOutput>,
```

Comment:
It is very strange to pass a mutable parameter to a function whose method name appears to be readonly (find_xxx), suggest to refactroing this fn params to immutable and return the `index` only:

```
fn find_acp_or_secp_belong_to_items<'a>(
        &self,
        cells: &'a Vec<packed::CellOutput>,
        items: &[Item],
    ) -> Option<usize> {
```

you may use the returned index value to get the mutable element:
```
change_to_existed_cell(&mut transfer_components.outputs[index], change_capacity);
```

Reply:
Personally, I think relocating element by index seems like a waste of time (not so elegant?). Anyway, there exists `find` method in rust whose arguments contain `&mut`, see https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find . As a result, I guess maybe returning the element we need is not a bad idea.

Reply:
@EthanYuan I think this PR is ready for merge, I will create a new PR to resolve this issue later.